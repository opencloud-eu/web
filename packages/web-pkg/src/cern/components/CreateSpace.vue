<template>
  <oc-button
    id="new-space-menu-btn"
    v-oc-tooltip="$gettext('Request new project')"
    type="a"
    href="https://cern.service-now.com/service-portal?id=sc_cat_item&name=request-storage-space&se=CERNBox-Service"
    target="_blank"
    appearance="filled"
  >
    <oc-icon name="add" />
    <span v-translate>New project</span>
  </oc-button>
</template>
