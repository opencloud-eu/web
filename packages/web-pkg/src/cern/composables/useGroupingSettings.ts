import { IncomingShareResource } from '@opencloud-eu/web-client'
import { computed, Ref, unref } from 'vue'

export const useGroupingSettings = ({
  sortBy,
  sortDir
}: {
  sortBy: Ref<string>
  sortDir: Ref<string>
}) => {
  const groupingSettings = computed(() => {
    return {
      groupingBy: localStorage.getItem('grouping-shared-with-me') || 'Shared on',
      showGroupingOptions: true,
      groupingFunctions: {
        'Name alphabetically': function (row: IncomingShareResource) {
          localStorage.setItem('grouping-shared-with-me', 'Name alphabetically')
          if (!isNaN(Number(row.name.charAt(0)))) {
            return '#'
          }
          if (row.name.charAt(0) === '.') {
            return row.name.charAt(1).toLowerCase()
          }
          return row.name.charAt(0).toLowerCase()
        },
        'Shared on': function (row: IncomingShareResource) {
          localStorage.setItem('grouping-shared-with-me', 'Shared on')
          const recently = Date.now() - 604800000
          const lastMonth = Date.now() - 2592000000
          if (Date.parse(row.sdate) < lastMonth) {
            return 'Older'
          }
          if (Date.parse(row.sdate) >= recently) {
            return 'Recently'
          } else {
            return 'Last month'
          }
        },
        'Share owner': function (row: IncomingShareResource) {
          localStorage.setItem('grouping-shared-with-me', 'Share owner')
          return row?.owner?.displayName
        },
        None: function () {
          localStorage.setItem('grouping-shared-with-me', 'None')
        }
      },
      sortGroups: {
        'Name alphabetically': function (groups: { name: string }[]) {
          // sort in alphabetical order by group name
          const sortedGroups = groups.sort(function (a, b) {
            if (a.name < b.name) {
              return -1
            }
            if (a.name > b.name) {
              return 1
            }
            return 0
          })
          // if sorting is done by name, reverse groups depending on asc/desc
          if (unref(sortBy) === 'name' && unref(sortDir) === 'desc') {
            sortedGroups.reverse()
          }
          return sortedGroups
        },
        'Shared on': function (groups: { name: string }[]) {
          // sort in order: 1-Recently, 2-Last month, 3-Older
          const sortedGroups = []
          const options = ['Recently', 'Last month', 'Older']
          for (const o of options) {
            const found = groups.find((el) => el.name.toLowerCase() === o.toLowerCase())
            if (found) {
              sortedGroups.push(found)
            }
          }
          // if sorting is done by sdate, reverse groups depending on asc/desc
          if (unref(sortBy) === 'sdate' && unref(sortDir) === 'asc') {
            sortedGroups.reverse()
          }
          return sortedGroups
        },
        'Share owner': function (groups: { name: string }[]) {
          // sort in alphabetical order by group name
          const sortedGroups = groups.sort(function (a, b) {
            if (a.name < b.name) {
              return -1
            }
            if (a.name > b.name) {
              return 1
            }
            return 0
          })
          // if sorting is done by owner, reverse groups depending on asc/desc
          if (unref(sortBy) === 'owner' && unref(sortDir) === 'desc') {
            sortedGroups.reverse()
          }
          return sortedGroups
        }
      }
    }
  })

  return {
    groupingSettings
  }
}
