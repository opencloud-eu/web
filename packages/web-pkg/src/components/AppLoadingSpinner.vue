<template>
  <div class="flex items-center justify-center oc-height-1-1 w-full">
    <oc-spinner id="app-loading-spinner" size="large" :aria-hidden="true" aria-label="" />
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'AppLoadingSpinner'
})
</script>
