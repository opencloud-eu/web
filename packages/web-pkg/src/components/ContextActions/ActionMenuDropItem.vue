<template>
  <li class="context-menu oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover">
    <oc-button
      :id="`oc-files-context-actions-${menuSection.name}-toggle`"
      appearance="raw"
      gap-size="medium"
      class="oc-drop-nested-toggle oc-py-s oc-px-m oc-width-1-1 oc-flex-justify-between oc-width-1-1"
    >
      <oc-icon :name="menuSection.drop.icon" size="medium" fill-type="line" />
      <span class="oc-flex oc-files-context-action-label">
        <span v-text="menuSection.drop.label" />
      </span>
      <oc-icon name="arrow-right-s" size="small" fill-type="line" />
    </oc-button>
    <oc-drop
      :toggle="`#oc-files-context-actions-${menuSection.name}-toggle`"
      :is-nested="true"
      mode="hover"
      class="oc-width-auto"
      padding-size="small"
      position="auto-start"
      close-on-click
    >
      <template v-if="menuSection.drop.items.length">
        <action-menu-item
          v-for="(action, actionIndex) in menuSection.drop.items"
          :key="`section-${menuSection.name}-action-${actionIndex}`"
          :action="action"
          :appearance="appearance"
          :action-options="actionOptions"
          class="oc-files-context-action oc-rounded oc-menu-item-hover"
        />
      </template>
      <span
        v-else
        class="oc-py-s oc-px-m oc-text-muted"
        v-text="menuSection.drop.emptyMessage || $gettext('No items available')"
      />
    </oc-drop>
  </li>
</template>

<script setup lang="ts">
import ActionMenuItem from './ActionMenuItem.vue'
import type { AppearanceType } from '@opencloud-eu/design-system/helpers'
import type { ActionOptions } from '../../composables'
import type { MenuSection } from './ContextActionMenu.vue'

const { menuSection, appearance, actionOptions } = defineProps<{
  menuSection: MenuSection
  appearance: AppearanceType
  actionOptions: ActionOptions
}>()
</script>
