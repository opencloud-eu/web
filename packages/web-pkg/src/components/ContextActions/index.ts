export { default as ActionMenuItem } from './ActionMenuItem.vue'
export { default as ContextActionMenu } from './ContextActionMenu.vue'
export { default as ContextMenuQuickAction } from './ContextMenuQuickAction.vue'
export { default as ActionMenuDropItem } from './ActionMenuDropItem.vue'
export * from './types'
