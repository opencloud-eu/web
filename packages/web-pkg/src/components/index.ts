export * from './AppBar'
export * from './AppTemplates'
export * from './ContextActions'
export * from './FilesList'
export * from './Filters'
export * from './Modals'
export * from './SideBar'
export * from './Search'
export * from './Spaces'
export * from './TextEditor'

export { default as AppLoadingSpinner } from './AppLoadingSpinner.vue'
export { default as AppTopBar } from './AppTopBar.vue'
export { default as BatchActions } from './BatchActions.vue'
export { default as CustomComponentTarget } from './CustomComponentTarget.vue'
export { default as ItemFilter } from './ItemFilter.vue'
export { default as ItemFilterToggle } from './ItemFilterToggle.vue'
export { default as LinkRoleDropdown } from './LinkRoleDropdown.vue'
export { default as LoadingIndicator } from './LoadingIndicator.vue'
export { default as NoContentMessage } from './NoContentMessage.vue'
export { default as Pagination } from './Pagination.vue'
export { default as QuotaSelect } from './QuotaSelect.vue'
export { default as SpaceQuota } from './SpaceQuota.vue'
export { default as SearchBarFilter } from './SearchBarFilter.vue'
export { default as ViewOptions } from './ViewOptions.vue'
export { default as PortalTarget } from './PortalTarget.vue'
export { default as CreateShortcutModal } from './CreateShortcutModal.vue'
export { default as CreateLinkModal } from './CreateLinkModal.vue'
