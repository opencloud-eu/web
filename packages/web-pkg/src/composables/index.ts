export * from './ability'
export * from './actions'
export * from './appDefaults'
export * from './appProviderService'
export * from './archiverService'
export * from './authContext'
export * from './breadcrumbs'
export * from './clientService'
export * from './clipboard'
export * from './download'
export * from './driveResolver'
export * from './embedMode'
export * from './eventBus'
export * from './fileListHeaderPosition'
export * from './filesList'
export * from './folderLink'
export * from './isTopBarSticky'
export * from './keyboardActions'
export * from './links'
export * from './loadingService'
export * from './localStorage'
export * from './pagination'
export * from './passwordPolicyService'
export * from './piniaStores'
export * from './portalTarget'
export * from './previewService'
export * from './requestHeaders'
export * from './resources'
export * from './router'
export * from './scrollTo'
export * from './search'
export * from './selection'
export * from './service'
export * from './shares'
export * from './sideBar'
export * from './sort'
export * from './spaces'
export * from './upload'
export * from './viewMode'
export * from './webWorkers'
export * from './avatars'
