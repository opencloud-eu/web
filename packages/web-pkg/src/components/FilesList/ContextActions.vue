<template>
  <context-action-menu
    :menu-sections="menuSections"
    :action-options="actionOptions"
    :drop-ref="dropRef"
  />
</template>

<script lang="ts">
import ContextActionMenu from '../ContextActions/ContextActionMenu.vue'
import { computed, defineComponent, PropType, Ref, toRef, unref } from 'vue'
import {
  ActionExtension,
  FileActionOptions,
  useExtensionRegistry,
  useFileActions,
  useFileActionsCopy,
  useFileActionsCopyPermanentLink,
  useFileActionsCreateSpaceFromResource,
  useFileActionsDelete,
  useFileActionsDisableSync,
  useFileActionsDownloadArchive,
  useFileActionsDownloadFile,
  useFileActionsEnableSync,
  useFileActionsFavorite,
  useFileActionsMove,
  useFileActionsNavigate,
  useFileActionsPaste,
  useFileActionsRename,
  useFileActionsRestore,
  useFileActionsSetImage,
  useFileActionsShowDetails,
  useFileActionsShowShares,
  useFileActionsToggleHideShare
} from '../../composables'
import { isNil } from 'lodash-es'
import { useGettext } from 'vue3-gettext'
import { MenuSection } from '../ContextActions'
import { OcDrop } from '@opencloud-eu/design-system/components'

export default defineComponent({
  name: 'ContextActions',
  components: { ContextActionMenu },
  props: {
    actionOptions: {
      type: Object as PropType<FileActionOptions>,
      required: true
    },
    dropRef: {
      type: Object as PropType<Ref<InstanceType<typeof OcDrop>>>,
      required: false,
      default: null
    }
  },
  setup(props) {
    const { editorActions, defaultEditorActions } = useFileActions()
    const { $gettext } = useGettext()

    const { actions: enableSyncActions } = useFileActionsEnableSync()
    const { actions: hideShareActions } = useFileActionsToggleHideShare()
    const { actions: copyActions } = useFileActionsCopy()
    const { actions: copyPermanentLinkActions } = useFileActionsCopyPermanentLink()
    const { actions: disableSyncActions } = useFileActionsDisableSync()
    const { actions: deleteActions } = useFileActionsDelete()
    const { actions: downloadArchiveActions } = useFileActionsDownloadArchive()
    const { actions: downloadFileActions } = useFileActionsDownloadFile()
    const { actions: favoriteActions } = useFileActionsFavorite()
    const { actions: moveActions } = useFileActionsMove()
    const { actions: navigateActions } = useFileActionsNavigate()
    const { actions: pasteActions } = useFileActionsPaste()
    const { actions: renameActions } = useFileActionsRename()
    const { actions: restoreActions } = useFileActionsRestore()
    const { actions: setSpaceImageActions } = useFileActionsSetImage()
    const { actions: showDetailsActions } = useFileActionsShowDetails()
    const { actions: createSpaceFromResourceActions } = useFileActionsCreateSpaceFromResource()
    const { actions: showSharesActions } = useFileActionsShowShares()

    const extensionRegistry = useExtensionRegistry()
    const extensionsContextActions = computed(() => {
      return extensionRegistry
        .requestExtensions<ActionExtension>({
          id: 'global.files.context-actions',
          extensionType: 'action'
        })
        .map((e) => e.action)
    })
    const extensionsBatchActions = computed(() => {
      return extensionRegistry
        .requestExtensions<ActionExtension>({
          id: 'global.files.batch-actions',
          extensionType: 'action'
        })
        .map((e) => e.action)
    })

    // type cast to make vue-tsc aware of the type
    const actionOptions = toRef(props, 'actionOptions') as Ref<FileActionOptions>

    const menuItemsBatchActions = computed(() =>
      [
        ...unref(enableSyncActions),
        ...unref(disableSyncActions),
        ...unref(downloadArchiveActions),
        ...unref(moveActions),
        ...unref(copyActions),
        ...unref(deleteActions),
        ...unref(restoreActions),
        ...unref(createSpaceFromResourceActions),
        ...unref(extensionsBatchActions).filter(
          (a) => a.category === 'actions' || isNil(a.category)
        )
      ].filter((item) => item.isVisible(unref(actionOptions)))
    )
    const menuItemsBatchSideBar = computed(() =>
      [
        ...unref(showDetailsActions),
        ...unref(extensionsBatchActions).filter((a) => a.category === 'sidebar')
      ].filter((item) => item.isVisible(unref(actionOptions)))
    )

    const menuItemsContext = computed(() => {
      return [...unref(navigateActions), ...unref(defaultEditorActions)]
        .filter((item) => item.isVisible(unref(actionOptions)))
        .sort((x, y) => Number(y.hasPriority) - Number(x.hasPriority))
    })

    const menuItemsContextDrop = computed(() => {
      return [
        ...unref(editorActions),
        ...unref(extensionsContextActions).filter((a) => a.category === 'context')
      ]
        .filter((item) => item.isVisible(unref(actionOptions)))
        .sort((x, y) => Number(y.hasPriority) - Number(x.hasPriority))
    })

    const menuItemsShare = computed(() => {
      return [
        ...unref(showSharesActions),
        ...unref(copyPermanentLinkActions),
        ...unref(extensionsContextActions).filter((a) => a.category === 'share')
      ].filter((item) => item.isVisible(unref(actionOptions)))
    })

    const menuItemsActions = computed(() => {
      return [
        ...unref(downloadArchiveActions),
        ...unref(downloadFileActions),
        ...unref(deleteActions),
        ...unref(moveActions),
        ...unref(copyActions),
        ...unref(pasteActions),
        ...unref(renameActions),
        ...unref(createSpaceFromResourceActions),
        ...unref(restoreActions),
        ...unref(enableSyncActions),
        ...unref(disableSyncActions),
        ...unref(hideShareActions),
        ...unref(setSpaceImageActions),
        ...unref(extensionsContextActions).filter(
          (a) => a.category === 'actions' || isNil(a.category)
        )
      ].filter((item) => item.isVisible(unref(actionOptions)))
    })

    const menuItemsSidebar = computed(() => {
      return [
        ...unref(favoriteActions).map((action) => {
          action.keepOpen = true
          return action
        }),
        ...unref(showDetailsActions),
        ...unref(extensionsContextActions).filter((a) => a.category === 'sidebar')
      ].filter((item) => item.isVisible(unref(actionOptions)))
    })

    const menuSections = computed(() => {
      const sections: MenuSection[] = []
      if (unref(actionOptions).resources.length > 1) {
        if (unref(menuItemsBatchActions).length) {
          sections.push({
            name: 'batch-actions',
            items: [...unref(menuItemsBatchActions)]
          })
        }

        sections.push({
          name: 'batch-details',
          items: [...unref(menuItemsBatchSideBar)]
        })
        return sections
      }

      if ([...unref(menuItemsContext), ...unref(menuItemsContextDrop)].length) {
        sections.push({
          name: 'context',
          items: [...unref(menuItemsContext)],
          dropItems: [
            {
              label: $gettext('Open with...'),
              name: 'open-with',
              icon: 'apps',
              items: [...unref(menuItemsContextDrop)]
            }
          ]
        })
      }

      if (unref(menuItemsShare).length) {
        sections.push({
          name: 'share',
          items: unref(menuItemsShare)
        })
      }
      if (unref(menuItemsActions).length) {
        sections.push({
          name: 'actions',
          items: unref(menuItemsActions)
        })
      }
      if (unref(menuItemsSidebar).length) {
        sections.push({
          name: 'sidebar',
          items: unref(menuItemsSidebar)
        })
      }
      return sections
    })

    return {
      menuSections
    }
  }
})
</script>
