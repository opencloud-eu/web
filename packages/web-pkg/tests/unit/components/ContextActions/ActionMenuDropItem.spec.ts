import ActionMenuDropItem from '../../../../src/components/ContextActions/ActionMenuDropItem.vue'
import { Action } from '../../../../src/composables/actions'
import { defaultPlugins, mount } from '@opencloud-eu/web-test-helpers'
import { MenuSectionDrop } from '../../../../src/components/ContextActions'

describe('ActionMenuDropItem component', () => {
  it('renders drop menu with actions', () => {
    const menuSectionDrop = {
      label: 'Actions',
      name: 'actions',
      icon: 'eye',
      items: [{ label: () => 'Copy' } as Action, { label: () => 'Paste' } as Action]
    }
    const { wrapper } = getWrapper(menuSectionDrop)
    expect(wrapper.html()).toMatchSnapshot()
    expect(
      wrapper.find('.oc-files-context-action-drop').findAll('.oc-files-context-action').length
    ).toEqual(menuSectionDrop.items.length)
  })
})

function getWrapper(menuSectionDrop: MenuSectionDrop) {
  return {
    wrapper: mount(ActionMenuDropItem, {
      props: {
        menuSectionDrop,
        actionOptions: { resources: [] },
        appearance: 'outline'
      },
      global: {
        plugins: [...defaultPlugins()]
      }
    })
  }
}
