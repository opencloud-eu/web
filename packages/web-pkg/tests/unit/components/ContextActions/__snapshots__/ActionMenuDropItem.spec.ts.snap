// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionMenuDropItem component > renders drop menu with actions 1`] = `
"<li class="context-menu oc-files-context-action oc-px-s oc-rounded oc-menu-item-hover"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw oc-py-s oc-px-m oc-width-1-1 oc-flex-justify-between oc-width-1-1" id="oc-files-context-actions-actions-toggle-2">
    <!--v-if-->
    <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m"><!----></span> <span class="oc-flex oc-files-context-action-label"><span>Actions</span></span> <span class="oc-icon oc-icon-s"><!----></span>
  </button>
  <div id="oc-files-context-actions-actions-drop-1" class="oc-drop oc-box-shadow-medium oc-rounded oc-width-auto oc-files-context-action-drop">
    <div class="oc-card oc-card-body oc-p-s">
<li class="oc-files-context-action oc-rounded oc-menu-item-hover"><button type="button" aria-label="" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-secondary oc-button-outline oc-button-secondary-outline action-menu-item oc-py-s oc-px-m oc-width-1-1" data-testid="action-handler">
    <!--v-if-->
    <!-- @slot Content of the button -->
    <!--v-if--> <span class="oc-files-context-action-label oc-flex" data-testid="action-label"><span>Copy</span></span>
    <!--v-if-->
  </button></li>
<li class="oc-files-context-action oc-rounded oc-menu-item-hover"><button type="button" aria-label="" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-m oc-button-secondary oc-button-outline oc-button-secondary-outline action-menu-item oc-py-s oc-px-m oc-width-1-1" data-testid="action-handler">
    <!--v-if-->
    <!-- @slot Content of the button -->
    <!--v-if--> <span class="oc-files-context-action-label oc-flex" data-testid="action-label"><span>Paste</span></span>
    <!--v-if-->
  </button></li>
</div>
</div>
</li>"
`;
